File: com.dexaint.dukechess/src/main/java/com/dexaint/dukechess/movement/Summon.java
package com.dexaint.dukechess.movement;

import org.apache.commons.lang3.ArrayUtils;

import com.dexaint.dukechess.flow.Field;
import com.dexaint.dukechess.flow.Player;

public class Summon extends MovementImpl {
	public int[] validateMovement(Field field, int pos, Destination d, Player p) {
		int[] ret = new int[]{};
		
		int[] offsets = d.dest2Offset(p);
		
		if (0 != offsets[0] & 0 != offsets[1] & Math.abs(offsets[0]) != Math.abs(offsets[1])) return ret;
		
		if (!(1 == Math.abs(offsets[0])) || (1 == Math.abs(offsets[1]))) return ret;
		
		for (int i = -1; i <= 1; i++)
		{
			for (int j = -1; j <= 1; j++)
			{
				int[] temp = {i, j};
				if(0 != Math.abs(temp[0]) + Math.abs(temp[1]))
				{
					ret = ArrayUtils.add(ret, offset2Dest(field, pos, d.dest2Offset(p)));
				}
			}
		}
		
		return ret;
	}
}

File: Unity/Assets/Scripts/Movement/Summon.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Summon : Movement
{
    public override List<int> ValidateMovement(List<ChessController> board, int pos, DestinationParser.Destination d, Player p)
    {
        List<int> ret = new List<int>();

        int[] offsets = DestinationParser.Dest2Offset(d, p);

        if (0 != offsets[0] & 0 != offsets[1] & Math.Abs(offsets[0]) != Math.Abs(offsets[1])) return ret;

        if (!(1 == Math.Abs(offsets[0]) || 1 == Math.Abs(offsets[1]))) return ret;

        if (!IsInField(pos, offsets) || HasAnyChess(board, pos, offsets)) return ret;

        ret.Add(Offset2Dest(pos, DestinationParser.Dest2Offset(d, p)));

        return ret;
    }
}
